java/ChartboostMediationDemoJava/src/main/java/com/example/chartboost/mediation/sdk/demo/java/MainActivity.java

```diff
@@ -0,0 +1,112 @@
+        setupLogsRecyclerView();
+
+        adController.initializeChartboost(
+                getString(R.string.default_app_id),
+                this,
+                getBaseContext(),
+                this
+        );
+
+        binding.loadInterstitialBtn.setOnClickListener(v -> onLoadClick(interstitialPlacementName, (Button) v, binding.showInterstitialBtn));
+        binding.showInterstitialBtn.setOnClickListener(v -> onShowClick(interstitialPlacementName, (Button) v));
+
+        binding.loadRewardedBtn.setOnClickListener(v -> onLoadClick(rewardedPlacementName, (Button) v, binding.showRewardedBtn));
+        binding.showRewardedBtn.setOnClickListener(v -> onShowClick(rewardedPlacementName, (Button) v));
+
+        binding.queueSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
+            adController.handleQueueToggle(isChecked);
+            addLog(String.format("Fullscreen Ad Queue will %s", isChecked ? "be used" : "not be used"));
+        });
+    }
+
+    private void setupLogsRecyclerView() {
+        logRecyclerView = binding.logsRv;
+        Button clearLogsButton = binding.clearLogsBtn;
+        logRecyclerView.setLayoutManager(new LinearLayoutManager(this));
+        logAdapter = new LogAdapter(uiLogs);
+        logRecyclerView.setAdapter(logAdapter);
+        clearLogsButton.setOnClickListener(v -> clearLogs());
+    }
+
+    private void onLoadClick(String placementName, Button loadButton, Button correspondingShowButton) {
+        loadButton.setEnabled(false);
+        adController.loadFullscreenAd(placementName, getBaseContext(), loadButton, correspondingShowButton, this);
+    }
+
+    private void onShowClick(String interstitialPlacementName, Button showButton) {
+        adController.showFullscreenAd(interstitialPlacementName, showButton, MainActivity.this, this);
+    }
+
+    private void loadBanner() {
+        adController.loadBanner(getString(R.string.banner_placement), this, binding.bannerLayout, getBaseContext());
+    }
+
+    private void addLog(String log) {
+        uiLogs.add(log);
+        logAdapter.notifyItemInserted(uiLogs.size() - 1);
+        logRecyclerView.post(() -> logRecyclerView.scrollToPosition(uiLogs.size() - 1));
+    }
+
+    private void clearLogs() {
+        uiLogs.clear();
+        logAdapter.notifyDataSetChanged();
+    }
+
+    @Override
+    public void add(String log) {
+        addLog(log);
+    }
+
+    @Override
+    public void onInitializationCompleted() {
+        loadBanner();
+        adController.createAdQueue(interstitialPlacementName, binding.showInterstitialBtn, getBaseContext(), this);
+        adController.createAdQueue(rewardedPlacementName, binding.showRewardedBtn, getBaseContext(), this);
+    }
+
+    @Override
+    public void onInitializationException() {
+        Toast.makeText(getBaseContext(), "Initialization failed", Toast.LENGTH_LONG).show();
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
```

Comment:
seconding this: all your loaded ads are not unloaded.
It's also not the greatest that you made the ad controller held by the activity. If the activity is destroyed, ads could still be loading in flight. also, the listener to the old ad queue is the old activity (or has access to the old activity).
It is possible to show ads loaded from another activity, so we should handle that (we don't need to showcase it; just make sure we're not leaking memory and doing general android best practices). It may not be the best idea to even unload the loaded ads, but if you don't have a good way to unload it, then that's fine; just destroy the loaded ad.
However, you also do have to handle the ad queue as that just keeps going. We should probably showcase to stop the ad queue when it's no longer used.